### PR TITLE
Idempotent Rest Publishing

### PR DESCRIPTION
**Pull request by ricardopereira**
#749

### CODE REVIEW

**Comment by ricardopereira** [comment]
`RSL1k4` and `RSL1k5` aren't passing, maybe because I should run the test suite with `dev` environment, I think.

**Comment by ricardopereira** [comment]
@paddybyers like you told me, `RTN10c` is failing in `dev`. Did you found anything suspicious?

**Comment by paddybyers** [comment]
Can you update and retry this pls? `dev` should now implement the spec fully

**Comment by ricardopereira** [comment]
@paddybyers still happens. Sorry.

```
Test Suite 'RealtimeClientConnection' started at 2018-09-06 09:39:16.419
Test Case '-[AblySpec.RealtimeClientConnection Connection__serial__should_have_last_known_connection_serial_from_restored_connection]' started.
/Users/ricardopereira/Repositories/Whitesmith/ably-ios/Spec/RealtimeClientConnection.swift:1441: error: -[AblySpec.RealtimeClientConnection Connection__serial__should_have_last_known_connection_serial_from_restored_connection] : Waited more than 10.0 seconds

Test Case '-[AblySpec.RealtimeClientConnection Connection__serial__should_have_last_known_connection_serial_from_restored_connection]' failed (11.735 seconds).
Test Suite 'RealtimeClientConnection' failed at 2018-09-06 09:39:28.156.
```

**Comment by paddybyers** [comment]
> still happens. Sorry.

@ricardopereira I can't find the log you created of the protocol traffic when this test runs. Can you re-send the link pls?

**Comment by ricardopereira** [comment]
@paddybyers Sure. https://gist.github.com/ricardopereira/642d8ec61958f51b21f836a58b72a85f

**Comment by ricardopereira** [comment]
@paddybyers Should I still use the `dev` environment?

**Comment by mattheworiordan** [comment]
@ricardopereira we will provide a stable env for you to use shortly.  @paddybyers please confirm when available.

**Comment by ricardopereira** [comment]
> @paddybyers like you told me, RTN10c is failing in dev. Did you found anything suspicious?

So after #800, `RTN10c` in `idempotent-dev` environment is still failing. Plus, a new test is failing in that environment: https://github.com/ably/ably-ios/blob/6a755c6fd52ea26be7d89043333968f86f558b9b/Spec/PushAdmin.swift#L232

```
✗ publish__should_publish_successfully, Waited more than 10.0 seconds
✗ Connection__serial__should_have_last_known_connection_serial_from_restored_connection, Waited more than 10.0 seconds
```

**Comment by mattheworiordan** [comment]
@ricardopereira I can't review for over a week. @paddybyers can you approve when you review? @ricardopereira you will need to make sure this PR disables default idempotency in 1.1 though, it's only enabled by default in 1.2+

File: Source/ARTClientOptions.m [hunk 1]
```diff
@@ -179,4 +181,13 @@ - (void)setDefaultTokenParams:(ARTTokenParams *)value {
     _defaultTokenParams = [[ARTTokenParams alloc] initWithTokenParams:value];
 }
 
++ (BOOL)getDefaultIdempotentRestPublishingForVersion:(NSString *)version {
+    if ([@"1.1" compare:version options:NSNumericSearch] == NSOrderedDescending) {
```
Comment (paddybyers) [inline Source/ARTClientOptions.m]: I think this now needs to change to 1.2

File: Spec/RestClientChannel.swift [hunk 2]
```diff
@@ -397,13 +397,295 @@ class RestClientChannel: QuickSpec {
...
+            }
+
+            // RSL1k
+            context("idempotent publishing") {
+
+                // TO3n
+                it("idempotentRestPublishing option") {
+                    expect(ARTClientOptions.getDefaultIdempotentRestPublishing(forVersion: "2")) == true
+                    expect(ARTClientOptions.getDefaultIdempotentRestPublishing(forVersion: "2.0.0")) == true
+                    expect(ARTClientOptions.getDefaultIdempotentRestPublishing(forVersion: "1.1")) == true
```
Comment (paddybyers) [inline Spec/RestClientChannel.swift]: This will be `false`

**Comment by ricardopereira** [comment]
@paddybyers `Connection__serial__should_have_last_known_connection_serial_from_restored_connection, Waited more than 10.0 seconds` is still failing. Should I mark it has pending and merge this?

**Pull request merged**